Follow up to PR31722 (#614)

ericcano opened the pull request:
#### PR description:

<!-- Please replace this text with a description of the feature proposed or problem addressed, specifying:
  - what changes are expected in the output if any, 
  - what other PRs or externals it depends upon if any,
  - link to any additional material useful to provide a documentation for this PR (slides, JIRA tickets, related pull requestes, hypernews, TWiki or Indico pages)  -->

#### PR validation:

<!-- Please replace this text with a description of which tests have been performed to verify the correctness of the PR, including the eventual addition of new code for testing like unit tests, test configurations, additions or updates to the runTheMatrix test workflows -->

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

<!-- Please replace this text with any link to  -->

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

ericcano committed:
Fixed include guards naming. (#614)

https://github.com/cms-sw/cmssw/pull/31722#pullrequestreview-623346747

ericcano committed:
Unified member access methods for class TrackSoAHeterogeneousT. (#614)

Used accessors in all cases. This commit does not change the memory layout of the class.
Also fixed C-style casts (reinterpret casts are needed here).
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603382606
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603383530

ericcano committed:
Fixed indlude guards naming. (#614)

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603384652

ericcano committed:
Fixed C-style cast. (#614)

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603386189

ericcano committed:
Reduced duplicate computations with temporary variables. (#614)

Also removed addition of transposed matrix by directly copying the needed elements.
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603395679
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603396371

ericcano committed:
Updated comments after methods renaming. (#614)

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603400548.

ericcano committed:
Modified PixelTrackProducer class to use esConsume(). (#614)

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603404617

ericcano committed:
Minor simplification (#614)

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603646172

fwyzard commented on CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h:
```diff
@@ -19,39 +19,61 @@ class TrackSoAHeterogeneousT {
+  constexpr Quality *qualityData() { return reinterpret_cast<Quality *>(quality_.data()); }
+
+  // chi2 accessors
+  __host__ __device__ auto & chi2(int32_t i) { return chi2_(i); }
```
question (I'm not sure myself): would t work to make these accessors `constexpr` ?

fwyzard replied:
also, could you make the return type explicit, instead of `auto` ?
I would like to be able to know if these return a `float` or a `double` or something else.

ericcano replied:
I think not. Quite the opposite, all the constexpr specifiers must be ignored by the compiler except for `stride()` and nHits(), which, by the way, are the same. I will simplify further,

fwyzard replied:
well, as long as the `constexpr` is valid, having it is a simpler shorthand for `__host__ __device__`

ericcano replied:
I missed that. This makes the file nice and homogeneous now.

fwyzard commented on CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h:
```diff
@@ -19,39 +19,61 @@ class TrackSoAHeterogeneousT {
+
+  // chi2 accessors
+  __host__ __device__ auto & chi2(int32_t i) { return chi2_(i); }
+  __host__ __device__ auto const chi2(int32_t i) const { return chi2_(i); }
```
if these are returned by value, there is no point/need for the `const` qualifier in the return type

ericcano replied:
Indeed, I will fix that.

fwyzard commented on CUDADataFormats/Track/interface/TrajectoryStateSoAT.h:
```diff
@@ -25,12 +25,12 @@ struct TrajectoryStateSoAT {
     cov(0) = ccov(0, 0);
     cov(1) = ccov(0, 1);
-    cov(2) = b * float(ccov(0, 2));
+    cov(2) = b * static_cast<float>(ccov(0, 2));
```
Fine by me - but I think that `float(...)` is a constructor, not a C-style cast.

ericcano committed:
Reduced computations with a local variable.

Also introduced a computation simplification suggested in:
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603648859

fwyzard commented on RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:
```diff
@@ -125,10 +125,13 @@ namespace riemannFit {
-    theta = theta < 0. ? theta + M_PI : theta;
+    const double p_t = std::min(20., fast_fit(2) * B);  // limit pt to avoid too small error!!!
+    // fast_fit(3) = tan(theta) =>
+    // 1 / sqr(sin(theta)) = (sqr(sin(theta) + sqr(cos(theta))) / sqr(sin(theta))
```
```suggestion
    // 1 / sqr(sin(theta)) = (sqr(sin(theta)) + sqr(cos(theta))) / sqr(sin(theta))
```

ericcano committed:
Update RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h

Co-authored-by: Andrea Bocci <andrea.bocci@cern.ch>

ericcano committed:
Simplified computations.

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603650560
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603653076
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603654860

VinInn commented on CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h:
```diff
@@ -19,39 +19,61 @@ class TrackSoAHeterogeneousT {
+
+  // stateAtBS accessors
+  constexpr auto & stateAtBS() { return stateAtBS_; }
+  constexpr auto stateAtBS() const { return stateAtBS_; }
```
are we sure to return it by value?

fwyzard replied:
You are right, probably not a good idea.

From a quick skim I thought these were scalars... for complex types, better return by `const &` (and please make the return type explicit, not `auto`).

ericcano replied:
Returning a const ref failed with this compilation error:
```
.../CMSSW_11_3_0_pre5/src/CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h:44:60: error: returning reference to temporary [-Werror=return-local-addr]
   44 |   constexpr auto const & pt(int32_t i) const { return pt_(i); }
```

It gets fixed by changing const eigen SoA operators like this:
```diff
diff --git a/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h b/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h
index 021402c93c2..186694bc1ec 100644
--- a/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h
+++ b/HeterogeneousCore/CUDAUtilities/interface/eigenSoA.h
@@ -19,9 +19,9 @@ namespace eigenSoA {
     using Scalar = T;

     __host__ __device__ constexpr Scalar& operator()(int32_t i) { return data_[i]; }
-    __device__ constexpr const Scalar operator()(int32_t i) const { return __ldg(data_ + i); }
+    __host__ __device__ constexpr Scalar const & operator()(int32_t i) const { return data_[i]; }
     __host__ __device__ constexpr Scalar& operator[](int32_t i) { return data_[i]; }
-    __device__ constexpr const Scalar operator[](int32_t i) const { return __ldg(data_ + i); }
+    __host__ __device__ constexpr Scalar const & operator[](int32_t i) const { return data_[i]; }

     __host__ __device__ constexpr Scalar* data() { return data_; }
     __host__ __device__ constexpr Scalar const* data() const { return data_; }
```
(which is more natural). Is this change fine too?

VinInn replied:
no is NOT!
please keep __ldg for device code. and return by value the scalar.

VinInn replied:
only ```stateAtBS_``` shall be returned by const ref. all others are scalar and shall be returned by value (for the very good reason you just discovered)

ericcano replied:
OK. I will just do the change with `stateAtBS` (as was requested). I thought the whole block needed the change.

ericcano committed:
Returned complex strcuture by reference instead of value.

Also made the type explicit.
https://github.com/cms-patatrack/cmssw/pull/615#discussion_r611021950

VinInn commented on CUDADataFormats/Track/interface/TrackSoAHeterogeneousT.h:
```diff
@@ -19,39 +19,61 @@ class TrackSoAHeterogeneousT {
-  HitContainer hitIndices;
-  HitContainer detIndices;
+  HitContainer hitIndices_;
+  HitContainer detIndices_;
```
I am not fully convinced of this encapsulations of SoAs.
It is just mere orthography. Semantics has not changed as both const and non-const accessors have been provided.
What are we encpsulating here?
The implementation encapsulation is done in the SoA itself.

Given that it is highly probable that all this will be modified again once the eventual final SoA model will be deployed (together with the Heterogeneous framework) I suggest to keep an open issue on Orthography of SoA
and leave for the time being my choice of orthography for my design of SoAs....

VinInn replied:
So please split the changes in the Arithmetic of the fit (that is supposed (I hope) a performance improvement) from this esthetic changes. (also because we want to profile them independently)

ericcano replied:
This commit just ensures we have a class with only one of each `public`, `protected` and `private` section. So indeed entirely cosmetic, but mandated by coding rules.

Computations and code style are already in different commits, so I guess you meant splitting this PR?

VinInn replied:
yes, spitting this PR:
one cosmetics
one computational performance

ericcano committed:
Fused class, header and SealModule.cc file.

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603659618

VinInn reviewed (commented):
We do this just because we are in this package, nothing to do with patatrack...
As usual we fix 20 year old code.

ericcano committed:
Removed unneeded range checking.

https://github.com/cms-sw/cmssw/pull/31722#discussion_r603666335
https://github.com/cms-sw/cmssw/pull/31722#discussion_r603666574

VinInn reviewed (commented):
One more fix that is not Patatrack related.
(in itself due since long time, still should be part of some unrelated cleanup campaign)

ericcano commented:
As requested by @VinInn, I close this PR to split it into a performance PR on one side and code cleanliness PR on the other.

VinInn commented:
I am now working on the pixel producers. I copied the version in here. Please do not commit them anymore.
